File: 6_zigzag_conversion/go/main.go
package main

import (
	"fmt"
	"strings"
)

func convert(s string, numRows int) string {
	if numRows <= 1 {
		return s
	}
	strs := make([]string, numRows, numRows)
	row, step := 0, 1
	for _, _s := range s {
		strs[row] += string(_s)
		if row == 0 {
			step = 1
		} else if (row == numRows - 1) {
			step = -1
		}
		row += step
	}

	return strings.Join(strs, "")
}

func main() {
	var s string

	s = "PAYPALISHIRING"
	fmt.Println(convert(s, 3))

	s = "AB"
	fmt.Println(convert(s, 1))
}

File: 6_zigzag_conversion/main.py
# -*- coding: utf-8 -*-


class Solution(object):

    def convert(self, s, numRows):
        if numRows <= 1:
            return s
        strs = [''] * numRows
        row, step = 0, 1
        for i in range(len(s)):
            strs[row] += s[i]
            if row == 0:
                step = 1
            elif row == numRows - 1:
                step = -1
            row += step

        return ''.join(strs)


if __name__ == '__main__':
    sl = Solution()

    s = 'PAYPALISHIRING'
    print(sl.convert(s, 3))

    s = 'AB'
    print(sl.convert(s, 1))
